aiida_restapi/graphql/nodes.py

```diff
@@ -0,0 +1,206 @@
+    fields_from_name,
+    multirow_cls_factory,
+    resolve_entity,
+    single_cls_factory,
+)
+from .utils import JSON, FilterString
+
+Link = type("LinkObjectType", (gr.ObjectType,), fields_from_name("Link"))
+
+
+class LinkQuery(gr.ObjectType):
+    """A link and its end node."""
+
+    link = gr.Field(Link)
+    # note: we must refer to this query using a string, to prevent circular dependencies
+    node = gr.Field("aiida_restapi.graphql.nodes.NodeQuery")
+
+
+class LinksQuery(multirow_cls_factory(LinkQuery, orm.nodes.Node, "nodes")):  # type: ignore[misc]
+    """Query all AiiDA Links."""
+
+
+class NodeQuery(
+    single_cls_factory(orm.nodes.Node, exclude_fields=("attributes", "extras"))  # type: ignore[misc]
+):
+    """Query an AiiDA Node"""
+
+    attributes = JSON(
+        description="Variable attributes of the node",
+        filter=gr.List(
+            gr.String,
+            description="return an exact set of attributes keys (non-existent will return null)",
+        ),
+    )
+    extras = JSON(
+        description="Variable extras (unsealed) of the node",
+        filter=gr.List(
+            gr.String,
+            description="return an exact set of extras keys (non-existent will return null)",
+        ),
+    )
+
+    # TODO it would be ideal if the attributes/extras were filtered via the SQL query
+
+    @staticmethod
+    def resolve_attributes(
+        parent: Any, info: gr.ResolveInfo, filter: Optional[List[str]] = None
+    ) -> Dict[str, Any]:
+        """Resolution function."""
+        attributes = parent.get("attributes")
+        if filter is None or attributes is None:
+            return attributes
+        return {key: attributes.get(key) for key in filter}
+
+    @staticmethod
+    def resolve_extras(
+        parent: Any, info: gr.ResolveInfo, filter: Optional[List[str]] = None
+    ) -> Dict[str, Any]:
+        """Resolution function."""
+        extras = parent.get("extras")
+        if filter is None or extras is None:
+            return extras
+        return {key: extras.get(key) for key in filter}
+
+    Comments = gr.Field(CommentsQuery, description="Comments attached to a node")
+
+    @staticmethod
+    def resolve_Comments(parent: Any, info: gr.ResolveInfo) -> dict:
+        """Resolution function."""
+        # pass filter specification to CommentsQuery
+        filters = {}
+        filters["dbnode_id"] = parent["id"]
+        return {"filters": filters}
+
+    Logs = gr.Field(LogsQuery, description="Logs attached to a process node")
+
+    @staticmethod
+    def resolve_Logs(parent: Any, info: gr.ResolveInfo) -> dict:
+        """Resolution function."""
+        # pass filter specification to CommentsQuery
+        filters = {}
+        filters["dbnode_id"] = parent["id"]
+        return {"filters": filters}
+
+    Incoming = gr.Field(
+        LinksQuery, description="Query for incoming nodes", filters=FilterString()
+    )
+
+    @staticmethod
+    def resolve_Incoming(
```

Comment:
so in graphene the `resolve_` methods have to be named exactly the same as the fields they are resolving. Since database fields, inherited from the SQL models, are lower case, I chose to distinguish "join" fields by capitlalizing them, e.g.

```
type NodeQuery {
  id: Int
  uuid: ID
  node_type: String
  process_type: String
  label: String
  description: String
  ctime: DateTime
  mtime: DateTime
  user_id: Int
  dbcomputer_id: Int
  attributes(filter: [String]): JSON
  extras(filter: [String]): JSON
  Comments: CommentsQuery
  Logs: LogsQuery
  Incoming(filters: FilterString): LinksQuery
  Outgoing(filters: FilterString): LinksQuery
  Ancestors(filters: FilterString): NodesQuery
  Descendants(filters: FilterString): NodesQuery
}
```

Reply:
Thanks for explaining the logic behind this choice.

I wonder whether this is intuitive from a user perspective though.
Users usually won't know the layout of the database and I think it will be difficult to guess whether a field should be upper or lower case. For example, I don't think a user of the API should have to know that `extras` are stored in the same table, while `Comments` are stored in a separate table (also, what if this should change at some point?).

I think a consistent naming scheme (all `snake_case` or all `camelCase`) would be more intuitive here.

By the way, the question of capitalization is actually somewhat independent of `snake_case` vs `camelCase` -in the  [graphene schema example](https://docs.graphene-python.org/en/latest/types/schema/#auto-camelcase-field-names) the fields of `Person` start with lower-case camel case; same in the [graphql docs](https://graphql.org/learn/schema/#interfaces).

What about
```
type NodeQuery {
  id: Int
  uuid: ID
  node_type: String
  process_type: String
  label: String
  description: String
  ctime: DateTime
  mtime: DateTime
  user_id: Int
  dbcomputer_id: Int
  attributes(filter: [String]): JSON
  extras(filter: [String]): JSON
  comments: CommentsQuery
  logs: LogsQuery
  incoming(filters: FilterString): LinksQuery
  outgoing(filters: FilterString): LinksQuery
  ancestors(filters: FilterString): NodesQuery
  descendants(filters: FilterString): NodesQuery
}
```

Reply:
hmm, I don't know if I agree here; its not just that the are stored in a different table, there is also a conceptual difference, that the "joins" are nested queries, i.e. you have to also provide fields for them and also often they directly relate to the top-level (standalone) queries, e.g.:

```graphql
Logs {
   count
   rows {
     id
   }
}
{
  Node {
    id
    Logs {
      count
      rows {
        id
      }
    }
  }
}
```

Reply:
essentially the uppercase "signals" that these fields require you to also supply nested fields, e.g. this will fail

```graphql
{
   Node {
     id
     Logs
   }
}
```

Reply:
in essence "lowecase == leaf node", "uppercase == tree node"

Reply:
> you have to also provide fields for them

Is it possible to define defaults here, e.g. default to returning all fields that aren't again joins?

By the way, my very limited browsing around in the web shows mostly lowercase being used also for top-level entities, e.g. https://blog.smartive.ch/graphql-and-mysql-solving-the-join-problem-191f40b55961
Are there any official guidelines on this?

Reply:
> Is it possible to define defaults here, e.g. default to returning all fields that aren't again joins?

I pretty sure not; in essence I think that goes against the design philosiphy of graphql, that you should request only exactly what you need

Reply:
> Are there any official guidelines on this?

Not to my knowledge;
I went for uppercase, because this directly mirrors the Python API, i.e. `from aiida orm import User, Node, Log, Comment, ...` and when you use the `Querybuilder`, e.g. `qb.append(User, project=['id'])` -> `User { id }` and so felt it would be most intuitive approach, and easiest to explain for those wishing to "translate" querybuilder code

Reply:
> so felt it would be most intuitive approach, and easiest to explain for those wishing to "translate" querybuilder code

I think users of the querybuilder will be used to `incomin`, `outgoing` etc. being lowercase.
I feel overall we will make our life easier if we just use lowercase throughout.

Also, I notice that the schema on https://blog.smartive.ch/graphql-and-mysql-solving-the-join-problem-191f40b55961  contains these nice elegant inline relationship specifications
```
 type Post {    id: ID    title: String    content: String    user: User! # to-one    comments: [Comment] # to-many  }  type Comment {    id: ID    content: String    post: Post! # to-one    user: User! # to-one  }  type User {    id: ID    name: String  }
  query {    post(id: ID): Post  }
```
Can we use that approach as well?

Reply:
> I feel overall we will make our life easier if we just use lowercase throughout.

Sure, I'm not too bothered, but wanted to explain my reasoning.

> contains these nice elegant inline relationship specifications

As in the `# to-many`, `# to-one` comments?
I'm literally just calling the `print_schema` method at present: https://github.com/aiidateam/aiida-restapi/blob/0d99673b2413acd827ed3df51b821a9c50e69612/aiida_restapi/graphql/sphinx_ext.py#L25
you would just need to implement your own aspects of the `graphql.utils.schema_printer` module to get whatever representation(s) you want. Obviously, when you use the GraphiQL interface, you get a lot more interactive feedback about the schema

Reply:
done (lowercase)